src/redux/actionCreators.ts

```diff
@@ -1078,57 +1085,31 @@ export const updateRouter = (
     previousAction: Action = null
 ) => (dispatch: DispatchWithState, getState: GetState) => {
     const { definition, components } = getState();
-    let newNode = { ...node };
-    const details = getDetails(newNode.uuid, components);
-    const previousNode = getNode(newNode.uuid, components, definition);
-    let newDef;
 
-    if (
-        details &&
-        !details.type &&
-        previousNode &&
-        previousNode.actions &&
-        previousNode.actions.length > 0
-    ) {
+    if (isActionSet(node.uuid, components, definition)) {
         // Make sure our previous action exists in our map
         if (previousAction && getDetails(previousAction.uuid, components)) {
-            return dispatch(spliceInRouter(newNode, type, previousAction));
+            dispatch(spliceInRouter(node, type, previousAction));
         } else {
-            return dispatch(appendNewRouter(newNode, type));
+            dispatch(appendNewRouter(node, type));
         }
-    }
-
-    if (draggedFrom) {
-        // console.log("adding new router node", props);
-        dispatch(
-            addNode(
-                newNode,
-                { position: newPosition, type },
-                {
-                    exitUUID: draggedFrom.exitUUID,
-                    nodeUUID: draggedFrom.nodeUUID
-                }
-            )
-        );
-        ({ freshestNode: newNode } = getState());
     } else {
-        // We're updating
-        ({ definition: newDef } = getState());
-        const nodeDetails = getDetails(newNode.uuid, components);
-        newDef = update(newDef, {
-            nodes: { [nodeDetails.nodeIdx]: { $set: newNode } }
-        });
-        newNode = newDef.nodes[nodeDetails.nodeIdx];
-        dispatch(updateNodeUI(newNode.uuid, { $merge: { type } }));
-        dispatch(setDefinition(newDef));
+        // Dragging from somewhere means we are a new node
```

Comment:
we could even make this more atomic/declarative if need be, e.g.

```
export const updateRouter = (
    node: Node,
    type: string,
    draggedFrom: DragPoint = null,
    newPosition: Position = null,
    previousAction: Action = null
) => (dispatch: DispatchWithState, getState: GetState) => {
    const { definition, components } = getState();

    if (isActionSet(node.uuid, components, definition)) {
        dispatch(routerFromAction(node, type, previousAction));
    } else {
        // Dragging from somewhere means we are a new node
        if (draggedFrom) {
            dispatch(addRouter(node, draggedFrom.exitUUID));
        } else {
            // Otherwise we are updating an existing node
            dispatch(
                applyUpdateSpec(prepSetNode(getDetails(node.uuid, components).nodeIdx, node, type))
            );
        }
    }

    console.timeEnd('updateRouter');
};
```

...`routerFromNode` and `addRouter` comprising the code in the `if (isAction(...` and `if (draggedfrom...` blocks